gamemodes/terrortown/entities/entities/ttt_logic_role.lua

```diff
@@ -35,24 +37,36 @@ function ENT:KeyValue(key, value)
 end
 
 ---
--- @param string name
--- @param Entity|Player activator
--- @return[default=true] boolean
--- @realm shared
-function ENT:AcceptInput(name, activator)
+-- Called when another entity fires an event to this entity.
+-- @param string name The name of the input that was triggered
+-- @param Entity|Player activator The initial cause for the input getting triggered (e.g. the player who pushed a button)
+-- @param Entity caller The entity that directly triggered the input (e.g. the button that was pushed)
+-- @param string data The data passed
+-- @return[default=true] boolean Return true if the default action should be supressed
+-- @realm server
+function ENT:AcceptInput(name, activator, caller, data)
+	local cv_evil_roles = GetConVar("ttt2_rolecheck_all_evil_roles")
+
 	if name == "TestActivator" then
-		if IsValid(activator) and activator:IsPlayer() then
-			local activator_role = (GetRoundState() == ROUND_PREP) and ROLE_INNOCENT or activator:GetBaseRole()
+		if not IsValid(activator) or not activator:IsPlayer() then return end
+
+		local activator_role = (GetRoundState() == ROUND_PREP) and ROLE_INNOCENT
+			---
+			-- @realm server
+			or roles.GetByIndex(hook.Run("TTT2ModifyLogicCheckRole", ply, self, activator, caller, data) or ply:GetSubRole()):GetBaseRole()
 
-			if self.Role == ROLE_NONE or self.Role == activator_role then
-				Dev(2, activator, "passed logic_role test of", self:GetName())
+		if self.Role == ROLE_TRAITOR
+			and (cv_evil_roles:GetBool() and (activator_role ~= ROLE_INNOCENT or activator_role ~= ROLE_DETECTIVE)
+				or not cv_evil_roles:GetBool() and (activator_role == ROLE_TRAITOR))
+			or self.Role == ROLE_NONE or self.Role == activator_role
```

Comment:
Maybe change the order. As it should check the Role None and activator role first, if they are  correct it works directly.
The cv_evil_roles is doubled. if the first is false, the second statement automatically is true, therefore it only needs no be activator_role == ROLE_TRAITOR.

Rest looks good.

Reply:
My fault, that is necessary in case the others are false, but evil_role isnt.....